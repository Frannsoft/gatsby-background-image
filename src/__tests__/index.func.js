import React from 'react'
import { render, fireEvent } from 'react-testing-library'
import 'react-testing-library/cleanup-after-each'
import './mocks/IntersectionObserver.mock'
import { mockAllIsIntersecting } from './mocks/IntersectionObserver.mock'

import {
  fixedShapeMock,
  fluidShapeMock,
  setupBackgroundImage,
} from './mocks/Various.mock'
import BackgroundImage from '../'
import {
  activateCacheForImage,
  createPictureRef,
  resetImageCache,
} from '../ImageUtils'

const LOAD_FAILURE_SRC = 'test_fluid_image.jpg'
const LOAD_SUCCESS_SRC = 'test_fixed_image.jpg'

describe(`<BackgroundImage /> with mock IO`, () => {
  const tmpImagePrototype = Object.getPrototypeOf(HTMLImageElement)

  beforeEach(() => {
    // Mocking HTMLImageElement.prototype.src to call the onload or onerror
    // callbacks depending on the src passed to it
    Object.defineProperty(HTMLImageElement.prototype, 'src', {
      // Define the property setter
      set(src) {
        this.setAttribute('src', src)
        if (src === LOAD_FAILURE_SRC) {
          // Call with setTimeout to simulate async loading
          setTimeout(() => this.onerror(new Error('mocked error')))
        } else if (src === LOAD_SUCCESS_SRC) {
          setTimeout(() => this.onload())
        }
      },
    })
    Object.defineProperty(HTMLImageElement.prototype, 'complete', {
      // Define the property setter
      set(complete) {
        this.complete = complete
      },
      get() {
        if (this.getAttribute('src') === LOAD_SUCCESS_SRC) {
          return true
        }
      },
    })

    resetImageCache()
  })
  afterEach(() => {
    Object.setPrototypeOf(HTMLImageElement, tmpImagePrototype)
  })

  it(`should render visible fluid image`, () => {
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642)
    let { container, rerender } = render(
      <BackgroundImage fixed={fixedShapeMock} />
    )
    mockAllIsIntersecting(true)
    expect(container).toMatchSnapshot()
    container = rerender(<BackgroundImage fluid={fluidShapeMock} />)
    expect(container).toMatchSnapshot()
  })

  it(`should render visible fixed image and call onLoadFunction`, () => {
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642)
    // Mock onStartLoad().
    const onLoadFunctionMock = jest.fn()
    let { container, rerender } = render(
      <BackgroundImage
        fluid={{ src: ``, aspectRatio: 1, srcSet: ``, sizes: `` }}
        onStartLoad={onLoadFunctionMock}
      />
    )
    mockAllIsIntersecting(true)
    expect(container).toMatchSnapshot()
    expect(onLoadFunctionMock).toHaveBeenCalled()
    container = rerender(<BackgroundImage fixed={fixedShapeMock} />)
    expect(container).toMatchSnapshot()
  })

  it(`should call critical fixed images`, () => {
    activateCacheForImage({ fixed: fixedShapeMock })
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
    }
    const component = setupBackgroundImage(options)
    mockAllIsIntersecting(true)
    expect(component).toMatchSnapshot()
  })

  it(`should not call onLoad without prop, fadeIn should stay falsy without seenBefore`, () => {
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
      onLoad: null,
    }
    const component = setupBackgroundImage(options)
    mockAllIsIntersecting(true)
    expect(component).toMatchSnapshot()
  })

  it(`should call onLoad and onError image events for LOAD_SUCCESS_SRC`, () => {
    const onLoadMock = jest.fn()
    const onErrorMock = jest.fn()
    const image = createPictureRef(
      {
        fixed: {
          src: LOAD_SUCCESS_SRC,
        },
        onLoad: onLoadMock,
        onError: onErrorMock,
      },
      onLoadMock
    )
    fireEvent.load(image)
    fireEvent.error(image)

    expect(onLoadMock).toHaveBeenCalledTimes(2)
    expect(onErrorMock).toHaveBeenCalledTimes(1)
  })

  it(`shouldn't call onLoad and onError image events without src`, () => {
    const onLoadMock = jest.fn()
    const onErrorMock = jest.fn()
    const image = createPictureRef(
      {
        fluid: {
          src: ``,
        },
        onLoad: onLoadMock,
        onError: onErrorMock,
      },
      onLoadMock
    )
    fireEvent.load(image)
    fireEvent.error(image)

    expect(onLoadMock).toHaveBeenCalledTimes(2)
    expect(onErrorMock).toHaveBeenCalledTimes(1)
  })
})

describe(`<BackgroundImage /> without IO`, () => {
  const tmpIO = global.IntersectionObserver
  beforeEach(() => {
    delete global.IntersectionObserver
    resetImageCache()
  })
  afterEach(() => {
    global.IntersectionObserver = tmpIO
  })

  it(`should call onLoadFunction without IO`, () => {
    const onLoadFunctionMock = jest.fn()
    activateCacheForImage({ fluid: fluidShapeMock })
    const options = {
      fluid: true,
      addClass: true,
      additionalClass: `test`,
      critical: true,
      onStartLoad: onLoadFunctionMock,
      fixed: false,
    }
    const component = setupBackgroundImage(options)
    expect(component).toMatchSnapshot()
    expect(onLoadFunctionMock).toHaveBeenCalled()
  })
})



