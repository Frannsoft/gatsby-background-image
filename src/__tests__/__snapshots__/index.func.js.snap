// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackgroundImage /> with mock IO should call critical fixed images 1`] = `
<div>
  <div
    class="imageClass gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline-block; width: 100px; height: 100px; background-size: 'contain';"
    title="Title for the image"
  >
    <style>
      
          .imageClass.:before,
.gatsby-background-image-test:before,
          .imageClass.:after,
.gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: 'center';
background-repeat: 'repeat-y';
background-size: 'contain';

          }
          .imageClass.:before,
.gatsby-background-image-test:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.:after,
.gatsby-background-image-test:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" /&gt;&lt;img width="100" height="100" srcset="some srcSet" src="test_fixed_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;z-index:-1;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should not call onLoad without prop, fadeIn should stay falsy without seenBefore 1`] = `
<div>
  <div
    class="imageClass gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline-block; width: 100px; height: 100px; background-size: 'contain';"
    title="Title for the image"
  >
    <style>
      
          .imageClass.:before,
.gatsby-background-image-test:before,
          .imageClass.:after,
.gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition-delay: 500ms;
            transition: opacity 0.5s;
            background-position: 'center';
background-repeat: 'repeat-y';
background-size: 'contain';

          }
          .imageClass.:before,
.gatsby-background-image-test:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.:after,
.gatsby-background-image-test:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" /&gt;&lt;img width="100" height="100" srcset="some srcSet" src="test_fixed_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;z-index:-1;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should render visible fixed image and call onLoadFunction 1`] = `
<div>
  <div
    class="gatsby-background-image-fwatluf_depr gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; background-position: center; background-repeat: no-repeat; background-size: cover;"
    title=""
  >
    <style>
      
          .gatsby-background-image-fwatluf_depr:before,
          .gatsby-background-image-fwatluf_depr:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition-delay: 0.25s;
            transition: opacity 0.5s;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gatsby-background-image-fwatluf_depr:before {
            z-index: -100;
            
            
            opacity: 1; 
          }
          .gatsby-background-image-fwatluf_depr:after {
            z-index: -101;
            
            
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;img src="" alt="" style="position:absolute;top:0;left:0;z-index:-1;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should render visible fixed image and call onLoadFunction 2`] = `undefined`;

exports[`<BackgroundImage /> with mock IO should render visible fluid image 1`] = `
<div>
  <div
    class="gatsby-background-image-fwatluf_depr gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; width: 100px; height: 100px; display: inline-block; background-position: center; background-repeat: no-repeat; background-size: cover;"
    title=""
  >
    <style>
      
          .gatsby-background-image-fwatluf_depr:before,
          .gatsby-background-image-fwatluf_depr:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition-delay: 0.25s;
            transition: opacity 0.5s;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gatsby-background-image-fwatluf_depr:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .gatsby-background-image-fwatluf_depr:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" /&gt;&lt;img width="100" height="100" srcset="some srcSet" src="test_fixed_image.jpg" alt="" style="position:absolute;top:0;left:0;z-index:-1;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should render visible fluid image 2`] = `undefined`;

exports[`<BackgroundImage /> without IO should call onLoadFunction without IO 1`] = `
<div>
  <div
    class="imageClass testgatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline; background-size: 'contain';"
    title="Title for the image"
  >
    <style>
      
          .imageClass.test:before,
.gatsby-background-image-test:before,
          .imageClass.test:after,
.gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition-delay: 0.25s;
            transition: opacity 0.5s;
            background-position: 'center';
background-repeat: 'repeat-y';
background-size: 'contain';

          }
          .imageClass.test:before,
.gatsby-background-image-test:before {
            z-index: -100;
            
            
            opacity: 0; 
          }
          .imageClass.test:after,
.gatsby-background-image-test:after {
            z-index: -101;
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;z-index:-1;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;
